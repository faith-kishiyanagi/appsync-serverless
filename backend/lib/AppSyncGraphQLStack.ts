import * as cdk from '@aws-cdk/core';
import * as appsync from '@aws-cdk/aws-appsync';
import * as lambda from '@aws-cdk/aws-lambda';

export class AppSyncGraphQLStack extends cdk.Stack {

  readonly api: appsync.GraphqlApi;

  constructor(scope: cdk.Construct, id: string, resolver: lambda.Function, props?: cdk.StackProps) {
    super(scope, id, props);

    this.api = new appsync.GraphqlApi(this, 'GqlApi', {
      name: 'blog-graphql-api',
      schema: appsync.Schema.fromAsset('graphql/schema.graphql'),
      authorizationConfig: {
        defaultAuthorization: {
          authorizationType: appsync.AuthorizationType.API_KEY,
          apiKeyConfig: {
            expires: cdk.Expiration.after(cdk.Duration.days(2)),
          }
        }
      }
    })

    const postResolver = this.api.addLambdaDataSource('PostResolver', resolver)

    postResolver.createResolver({
      typeName: 'Query',
      fieldName: 'allPosts'
    });
    postResolver.createResolver({
      typeName: 'Query',
      fieldName: 'findOnePostBy'
    });
    postResolver.createResolver({
      typeName: 'Mutation',
      fieldName: 'createPost'
    });
    postResolver.createResolver({
      typeName: 'Mutation',
      fieldName: 'updatePost'
    });
    postResolver.createResolver({
      typeName: 'Mutation',
      fieldName: 'deletePost'
    });

    new cdk.CfnOutput(this, 'AppSyncUrl', {
      value: this.api.graphqlUrl
    });
    new cdk.CfnOutput(this, 'AppSyncApiKey', {
      value: this.api.apiKey || ''
    });
    new cdk.CfnOutput(this, 'ProjectRegion', {
      value: this.region
    });
  }
}
