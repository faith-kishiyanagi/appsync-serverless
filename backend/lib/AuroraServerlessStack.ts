import * as cdk from '@aws-cdk/core';
import * as ec2 from '@aws-cdk/aws-ec2';
import * as rds from '@aws-cdk/aws-rds';
import { AuroraCapacityUnit } from '@aws-cdk/aws-rds';

export class AuroraServerlessStack extends cdk.Stack {

  readonly vpc: ec2.Vpc;
  readonly blogDB: rds.ServerlessCluster;

  constructor(scope: cdk.Construct, id: string, props?: cdk.StackProps) {
    super(scope, id, props);

    this.vpc = new ec2.Vpc(this, 'BlogAppVpc', {
      cidr: '10.0.1.0/24',
      natGateways: 0,
      subnetConfiguration: [
        {cidrMask: 26, name: 'Isolated', subnetType: ec2.SubnetType.ISOLATED}
      ],
    })

    this.blogDB = new rds.ServerlessCluster(this, 'BlogDbAuroraServerlessCluster', {
      engine: rds.DatabaseClusterEngine.AURORA_MYSQL,
      parameterGroup: rds.ParameterGroup.fromParameterGroupName(this, 'BlogDbParameterGroup', 'default.aurora-mysql5.7'),
      defaultDatabaseName: 'BlogDB',
      vpc: this.vpc,
      vpcSubnets: {subnetType: ec2.SubnetType.ISOLATED},
      scaling: {
        autoPause: cdk.Duration.minutes(5),
        minCapacity: AuroraCapacityUnit.ACU_1,
        maxCapacity: AuroraCapacityUnit.ACU_2,
      }
    })
  }
}
