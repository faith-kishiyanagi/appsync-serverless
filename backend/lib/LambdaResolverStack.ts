import * as cdk from '@aws-cdk/core';
import * as lambda from '@aws-cdk/aws-lambda';
import * as rds from '@aws-cdk/aws-rds';

export class LambdaResolverStack extends cdk.Stack {

  readonly postFunction: lambda.Function;

  constructor(scope: cdk.Construct, id: string, db: rds.ServerlessCluster, props?: cdk.StackProps) {
    super(scope, id, props);

    this.postFunction = new lambda.Function(this, 'PostFunction', {
      runtime: lambda.Runtime.NODEJS_14_X,
      code: new lambda.AssetCode('functions/post'),
      handler: 'src/index.handler',
      memorySize: 256,
      timeout: cdk.Duration.seconds(30),
      environment: {
        CLUSTER_ARN: db.clusterArn,
        SECRET_ARN: db.secret?.secretArn || '',
        DB_NAME: 'BlogDB',
        AWS_NODEJS_CONNECTION_REUSE_ENABLED: '1'
      }
    })

    db.grantDataApiAccess(this.postFunction);
  }
}
