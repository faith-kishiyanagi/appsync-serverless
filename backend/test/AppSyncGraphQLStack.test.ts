import '@aws-cdk/assert/jest'
import { SynthUtils } from '@aws-cdk/assert';
import * as cdk from '@aws-cdk/core';
import { AppSyncGraphQLStack } from '../lib/AppSyncGraphQLStack';
import { LambdaResolverStack } from '../lib/LambdaResolverStack';
import { AuroraServerlessStack } from "../lib/AuroraServerlessStack";

const app = new cdk.App();
const aurora = new AuroraServerlessStack(app, 'AuroraServerless');
const lambda = new LambdaResolverStack(app, 'LambdaResolver', aurora.blogDB)
const stack = new AppSyncGraphQLStack(app, 'AppSyncGraphQL', lambda.postFunction);

test('Stack snapshot', () => {
  expect(SynthUtils.toCloudFormation(stack)).toMatchSnapshot();
});

test('Configured AppSync GraphQL API', () => {
  expect(stack).toHaveResource('AWS::AppSync::GraphQLApi', {
    'Name': 'blog-graphql-api',
    'AuthenticationType': 'API_KEY',
  });
});
