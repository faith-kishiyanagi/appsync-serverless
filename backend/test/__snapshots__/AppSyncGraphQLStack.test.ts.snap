// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Stack snapshot 1`] = `
Object {
  "Outputs": Object {
    "AppSyncApiKey": Object {
      "Value": Object {
        "Fn::GetAtt": Array [
          "GqlApiDefaultApiKey3A4E3881",
          "ApiKey",
        ],
      },
    },
    "AppSyncUrl": Object {
      "Value": Object {
        "Fn::GetAtt": Array [
          "GqlApi4E487465",
          "GraphQLUrl",
        ],
      },
    },
    "ProjectRegion": Object {
      "Value": Object {
        "Ref": "AWS::Region",
      },
    },
  },
  "Resources": Object {
    "GqlApi4E487465": Object {
      "Properties": Object {
        "AuthenticationType": "API_KEY",
        "Name": "blog-graphql-api",
      },
      "Type": "AWS::AppSync::GraphQLApi",
    },
    "GqlApiDefaultApiKey3A4E3881": Object {
      "DependsOn": Array [
        "GqlApiSchema55BCA17B",
      ],
      "Properties": Object {
        "ApiId": Object {
          "Fn::GetAtt": Array [
            "GqlApi4E487465",
            "ApiId",
          ],
        },
        "Expires": 1617268006,
      },
      "Type": "AWS::AppSync::ApiKey",
    },
    "GqlApiPostResolverCEF0036B": Object {
      "Properties": Object {
        "ApiId": Object {
          "Fn::GetAtt": Array [
            "GqlApi4E487465",
            "ApiId",
          ],
        },
        "LambdaConfig": Object {
          "LambdaFunctionArn": Object {
            "Fn::ImportValue": "LambdaResolver:ExportsOutputFnGetAttPostFunction61046647Arn50D23A83",
          },
        },
        "Name": "PostResolver",
        "ServiceRoleArn": Object {
          "Fn::GetAtt": Array [
            "GqlApiPostResolverServiceRoleA926AB19",
            "Arn",
          ],
        },
        "Type": "AWS_LAMBDA",
      },
      "Type": "AWS::AppSync::DataSource",
    },
    "GqlApiPostResolverMutationcreatePostResolver78B052AC": Object {
      "DependsOn": Array [
        "GqlApiPostResolverCEF0036B",
        "GqlApiSchema55BCA17B",
      ],
      "Properties": Object {
        "ApiId": Object {
          "Fn::GetAtt": Array [
            "GqlApi4E487465",
            "ApiId",
          ],
        },
        "DataSourceName": "PostResolver",
        "FieldName": "createPost",
        "Kind": "UNIT",
        "TypeName": "Mutation",
      },
      "Type": "AWS::AppSync::Resolver",
    },
    "GqlApiPostResolverMutationdeletePostResolverA00B368C": Object {
      "DependsOn": Array [
        "GqlApiPostResolverCEF0036B",
        "GqlApiSchema55BCA17B",
      ],
      "Properties": Object {
        "ApiId": Object {
          "Fn::GetAtt": Array [
            "GqlApi4E487465",
            "ApiId",
          ],
        },
        "DataSourceName": "PostResolver",
        "FieldName": "deletePost",
        "Kind": "UNIT",
        "TypeName": "Mutation",
      },
      "Type": "AWS::AppSync::Resolver",
    },
    "GqlApiPostResolverMutationupdatePostResolverE032C7A3": Object {
      "DependsOn": Array [
        "GqlApiPostResolverCEF0036B",
        "GqlApiSchema55BCA17B",
      ],
      "Properties": Object {
        "ApiId": Object {
          "Fn::GetAtt": Array [
            "GqlApi4E487465",
            "ApiId",
          ],
        },
        "DataSourceName": "PostResolver",
        "FieldName": "updatePost",
        "Kind": "UNIT",
        "TypeName": "Mutation",
      },
      "Type": "AWS::AppSync::Resolver",
    },
    "GqlApiPostResolverQueryallPostsResolver63E549DD": Object {
      "DependsOn": Array [
        "GqlApiPostResolverCEF0036B",
        "GqlApiSchema55BCA17B",
      ],
      "Properties": Object {
        "ApiId": Object {
          "Fn::GetAtt": Array [
            "GqlApi4E487465",
            "ApiId",
          ],
        },
        "DataSourceName": "PostResolver",
        "FieldName": "allPosts",
        "Kind": "UNIT",
        "TypeName": "Query",
      },
      "Type": "AWS::AppSync::Resolver",
    },
    "GqlApiPostResolverQueryfindOnePostByResolver60EB5B03": Object {
      "DependsOn": Array [
        "GqlApiPostResolverCEF0036B",
        "GqlApiSchema55BCA17B",
      ],
      "Properties": Object {
        "ApiId": Object {
          "Fn::GetAtt": Array [
            "GqlApi4E487465",
            "ApiId",
          ],
        },
        "DataSourceName": "PostResolver",
        "FieldName": "findOnePostBy",
        "Kind": "UNIT",
        "TypeName": "Query",
      },
      "Type": "AWS::AppSync::Resolver",
    },
    "GqlApiPostResolverServiceRoleA926AB19": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "appsync.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::IAM::Role",
    },
    "GqlApiPostResolverServiceRoleDefaultPolicy9CA016F8": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "lambda:InvokeFunction",
              "Effect": "Allow",
              "Resource": Object {
                "Fn::ImportValue": "LambdaResolver:ExportsOutputFnGetAttPostFunction61046647Arn50D23A83",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "GqlApiPostResolverServiceRoleDefaultPolicy9CA016F8",
        "Roles": Array [
          Object {
            "Ref": "GqlApiPostResolverServiceRoleA926AB19",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "GqlApiSchema55BCA17B": Object {
      "Properties": Object {
        "ApiId": Object {
          "Fn::GetAtt": Array [
            "GqlApi4E487465",
            "ApiId",
          ],
        },
        "Definition": "type Post {
    id: String!
    title: String!
    content: String!
}

input CreatePostInput {
    title: String!
    content: String!
}

input UpdatePostInput {
    id: String!
    title: String
    content: String
}

type Query {
    allPosts: [Post]
    findOnePostBy(postId: String!): Post
}

type Mutation {
    createPost(post: CreatePostInput): Post
    updatePost(post: UpdatePostInput): Post
    deletePost(postId: String!): String
}

type Subscription {
    onCreatePost: Post
        @aws_subscribe(mutations: [\\"createPost\\"])
    onUpdatePost: Post
        @aws_subscribe(mutations: [\\"updatePost\\"])
    onDeletePost: String
        @aws_subscribe(mutations: [\\"deletePost\\"])
}
",
      },
      "Type": "AWS::AppSync::GraphQLSchema",
    },
  },
}
`;
